File: refund-exists-but-the-money-never-moved/node/find-stuck-refunds.js
/**
 * Flag Shopify orders whose refund exists but the money never actually moved.
 *
 * A refund record can sit on an order while the gateway transaction underneath it
 * failed, errored, or never left PENDING. This sums each refund's SUCCESS-only
 * transactions in minor units, compares that to the refund's own totalRefundedSet,
 * and tags the order for review with tagsAdd when they do not match. Run on a
 * schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/shopify/refund-exists-but-the-money-never-moved/
 */
import { pathToFileURL } from "node:url";

const SHOP = process.env.SHOPIFY_SHOP || "example.myshopify.com";
const TOKEN = process.env.SHOPIFY_ACCESS_TOKEN || "shpat_dummy";
const API_VERSION = process.env.SHOPIFY_API_VERSION || "2025-01";
const ENDPOINT = `https://${SHOP}/admin/api/${API_VERSION}/graphql.json`;
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 14);
const REVIEW_TAG = process.env.REVIEW_TAG || "refund-stuck";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function toCents(amount) {
  return Math.round(parseFloat(amount) * 100);
}

export function movedCents(refund) {
  let total = 0;
  for (const t of refund.transactions?.nodes || []) {
    if (t.status === "SUCCESS") total += toCents(t.amountSet.shopMoney.amount);
  }
  return total;
}

export function isStuckRefund(refund) {
  const claimed = toCents(refund.totalRefundedSet?.shopMoney?.amount ?? "0");
  return Math.abs(movedCents(refund) - claimed) > 1;
}

export function hasStuckRefund(order) {
  return (order.refunds || []).some(isStuckRefund);
}

async function gql(query, variables = {}) {
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: { "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Shopify ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const ORDERS_QUERY = `
query($cursor: String, $q: String!) {
  orders(first: 25, after: $cursor, query: $q) {
    pageInfo { hasNextPage endCursor }
    nodes {
      id name tags
      refunds {
        id
        totalRefundedSet { shopMoney { amount currencyCode } }
        transactions(first: 10) {
          nodes { kind status amountSet { shopMoney { amount currencyCode } } }
        }
      }
    }
  }
}`;

const TAGS_ADD = `
mutation($id: ID!, $tags: [String!]!) {
  tagsAdd(id: $id, tags: $tags) { node { id } userErrors { field message } }
}`;

async function* refundedOrders() {
  const q = `created_at:>-${LOOKBACK_DAYS}d AND financial_status:refunded OR financial_status:partially_refunded`;
  let cursor = null;
  while (true) {
    const data = (await gql(ORDERS_QUERY, { cursor, q })).orders;
    for (const node of data.nodes) yield node;
    if (!data.pageInfo.hasNextPage) return;
    cursor = data.pageInfo.endCursor;
  }
}

async function tagForReview(orderId, reviewTag) {
  const result = (await gql(TAGS_ADD, { id: orderId, tags: [reviewTag] })).tagsAdd;
  if (result.userErrors.length) throw new Error(JSON.stringify(result.userErrors));
}

export async function run() {
  let flagged = 0;
  for await (const order of refundedOrders()) {
    if (!hasStuckRefund(order)) continue;
    if ((order.tags || []).includes(REVIEW_TAG)) continue;
    console.warn(`Order ${order.name} has a refund that never moved the money. ${DRY_RUN ? "would tag" : "tagging"}`);
    if (!DRY_RUN) await tagForReview(order.id, REVIEW_TAG);
    flagged++;
  }
  console.log(`Done. ${flagged} order(s) ${DRY_RUN ? "to tag" : "tagged"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: refund-exists-but-the-money-never-moved/python/find_stuck_refunds.py
"""Flag Shopify orders whose refund exists but the money never actually moved.

A refund record can sit on an order while the gateway transaction underneath it
failed, errored, or never left PENDING. This sums each refund's SUCCESS-only
transactions in minor units, compares that to the refund's own totalRefundedSet,
and tags the order for review with tagsAdd when they do not match. Read only
apart from the tag. Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/shopify/refund-exists-but-the-money-never-moved/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("find_stuck_refunds")

SHOP = os.environ.get("SHOPIFY_SHOP", "example.myshopify.com")
TOKEN = os.environ.get("SHOPIFY_ACCESS_TOKEN", "shpat_dummy")
API_VERSION = os.environ.get("SHOPIFY_API_VERSION", "2025-01")
ENDPOINT = f"https://{SHOP}/admin/api/{API_VERSION}/graphql.json"
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", "14"))
REVIEW_TAG = os.environ.get("REVIEW_TAG", "refund-stuck")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ORDERS_QUERY = """
query($cursor: String, $q: String!) {
  orders(first: 25, after: $cursor, query: $q) {
    pageInfo { hasNextPage endCursor }
    nodes {
      id name tags
      refunds {
        id
        totalRefundedSet { shopMoney { amount currencyCode } }
        transactions(first: 10) {
          nodes { kind status amountSet { shopMoney { amount currencyCode } } }
        }
      }
    }
  }
}"""

TAGS_ADD = """
mutation($id: ID!, $tags: [String!]!) {
  tagsAdd(id: $id, tags: $tags) { node { id } userErrors { field message } }
}"""


def gql(query, variables=None):
    r = requests.post(
        ENDPOINT,
        json={"query": query, "variables": variables or {}},
        headers={"X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def to_cents(amount):
    return round(float(amount) * 100)


def moved_cents(refund):
    """Sum of only the SUCCESS transactions on a refund, in minor units."""
    total = 0
    for t in (refund.get("transactions") or {}).get("nodes", []):
        if t.get("status") == "SUCCESS":
            total += to_cents(t["amountSet"]["shopMoney"]["amount"])
    return total


def is_stuck_refund(refund):
    claimed = to_cents(refund.get("totalRefundedSet", {}).get("shopMoney", {}).get("amount", "0"))
    return abs(moved_cents(refund) - claimed) > 1


def has_stuck_refund(order):
    return any(is_stuck_refund(r) for r in order.get("refunds") or [])


def tag_for_review(order_id, review_tag):
    result = gql(TAGS_ADD, {"id": order_id, "tags": [review_tag]})["tagsAdd"]
    if result["userErrors"]:
        raise RuntimeError(result["userErrors"])


def refunded_orders():
    q = f"created_at:>-{LOOKBACK_DAYS}d AND financial_status:refunded OR financial_status:partially_refunded"
    cursor = None
    while True:
        data = gql(ORDERS_QUERY, {"cursor": cursor, "q": q})["orders"]
        for node in data["nodes"]:
            yield node
        if not data["pageInfo"]["hasNextPage"]:
            return
        cursor = data["pageInfo"]["endCursor"]


def run():
    flagged = 0
    for order in refunded_orders():
        if not has_stuck_refund(order):
            continue
        if REVIEW_TAG in (order.get("tags") or []):
            continue
        log.warning("Order %s has a refund that never moved the money. %s",
                    order["name"], "would tag" if DRY_RUN else "tagging")
        if not DRY_RUN:
            tag_for_review(order["id"], REVIEW_TAG)
        flagged += 1
    log.info("Done. %d order(s) %s.", flagged, "to tag" if DRY_RUN else "tagged")


if __name__ == "__main__":
    run()
